pkg/strategy/factorzoo/correlation.go

```diff
@@ -0,0 +1,103 @@
+package factorzoo
+
+import (
+	"fmt"
+	"github.com/c9s/bbgo/pkg/indicator"
+	"github.com/c9s/bbgo/pkg/types"
+	"math"
+	"time"
+)
+
+var zeroTime time.Time
+
+//var zeroTime time.Time
+type KLineValueMapper func(k types.KLine) float64
+
+//go:generate callbackgen -type Correlation
+type Correlation struct {
+	types.IntervalWindow
+	Values  types.Float64Slice
+	EndTime time.Time
+
+	UpdateCallbacks []func(value float64)
+}
+
+func (inc *Correlation) Last() float64 {
+	if len(inc.Values) == 0 {
+		return 0.0
+	}
+	return inc.Values[len(inc.Values)-1]
+}
+
+func (inc *Correlation) calculateAndUpdate(klines []types.KLine) {
+	if len(klines) < inc.Window {
+		return
+	}
+
+	var end = len(klines) - 1
+	var lastKLine = klines[end]
+
+	if inc.EndTime != zeroTime && lastKLine.GetEndTime().Before(inc.EndTime) {
+		return
+	}
+
+	var recentT = klines[end-(inc.Window-1) : end+1]
+
+	correlation, err := calculateCORRELATION(recentT, inc.Window, KLineAmplitudeMapper, indicator.KLineVolumeMapper)
+	if err != nil {
+		log.WithError(err).Error("can not calculate correlation")
+		return
+	}
+	inc.Values.Push(correlation)
+
+	if len(inc.Values) > indicator.MaxNumOfVOL {
+		inc.Values = inc.Values[indicator.MaxNumOfVOLTruncateSize-1:]
+	}
+
+	inc.EndTime = klines[end].GetEndTime().Time()
+
+	inc.EmitUpdate(correlation)
+}
+
+func (inc *Correlation) handleKLineWindowUpdate(interval types.Interval, window types.KLineWindow) {
+	if inc.Interval != interval {
+		return
+	}
+
+	inc.calculateAndUpdate(window)
+}
+
+func (inc *Correlation) Bind(updater indicator.KLineWindowUpdater) {
+	updater.OnKLineWindowUpdate(inc.handleKLineWindowUpdate)
+}
+
+func calculateCORRELATION(klines []types.KLine, window int, valA KLineValueMapper, valB KLineValueMapper) (float64, error) {
+	length := len(klines)
+	if length == 0 || length < window {
+		return 0.0, fmt.Errorf("insufficient elements for calculating VOL with window = %d", window)
+	}
+
+	sumA, sumB, sumAB, squareSumA, squareSumB := 0., 0., 0., 0., 0.
+	for _, k := range klines {
+		// sum of elements of array A
+		sumA += valA(k)
+		// sum of elements of array B
+		sumB += valA(k)
+
+		// sum of A[i] * B[i].
+		sumAB = sumAB + valA(k)*valB(k)
+
+		// sum of square of array elements.
+		squareSumA = squareSumA + valA(k)*valA(k)
+		squareSumB = squareSumB + valB(k)*valB(k)
+	}
```

Comment:
How about add a `toFloatSlice` function which maps klines to `types.Float64Slice`, and calculate sums by `Sum` and `Dot`?

```
sliceA := toFloatSlice(klines, valA)
sliceB := toFloatSlice(klines, valB)

sumA := sliceA.Sum()
sumB := sliceB.Sum()

sumAB := sliceA.Dot(sliceB)

squareSumA := sliceA.Dot(sliceA)
squareSumB := sliceB.Dot(sliceB)
```

Reply:
or 
https://pkg.go.dev/gonum.org/v1/gonum/stat#Correlation

```
sliceA := toFloatSlice(klines, valA) sliceB := toFloatSlice(klines, valB)
ones := types.NewOneFloat64Slice(len(sliceA))
corr := stat.Correlation(sliceA, sliceB, ones)
```